tests/unit/s2n_self_talk_session_resumption_test.c

```diff
@@ -441,6 +445,92 @@ int main(int argc, char **argv)
+
+        DEFER_CLEANUP(struct s2n_stuffer tls12_ticket = { 0 }, s2n_stuffer_free);
+        DEFER_CLEANUP(struct s2n_stuffer tls13_ticket = { 0 }, s2n_stuffer_free);
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&tls12_ticket, 0));
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&tls13_ticket, 0));
+
+        /* Negotiate initial TLS1.3 handshake */
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, tls13_client_config));
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_TRUE(ARE_FULL_HANDSHAKES(client_conn, server_conn));
+        EXPECT_OK(s2n_test_issue_new_session_ticket(server_conn, client_conn, &no_early_data));
+        EXPECT_SUCCESS(s2n_stuffer_copy(&cb_session_data, &tls13_ticket, s2n_stuffer_data_available(&cb_session_data)));
+
+        /* Prepare client and server for a second connection */
+        EXPECT_SUCCESS(s2n_connection_wipe(client_conn));
+        EXPECT_SUCCESS(s2n_connection_wipe(server_conn));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        /* Negotiate initial TLS1.2 handshake */
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, tls12_client_config));
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_TRUE(ARE_FULL_HANDSHAKES(client_conn, server_conn));
+        int tls12_ticket_length = s2n_connection_get_session_length(client_conn);
+        EXPECT_SUCCESS(s2n_stuffer_skip_write(&tls12_ticket, tls12_ticket_length));
+        EXPECT_SUCCESS(s2n_connection_get_session(client_conn, tls12_ticket.blob.data, tls12_ticket_length));
+
+        /* Switch between TLS1.2 and TLS1.3 resumption */
+        for (size_t i = 0; i < 10; i++) {
+            EXPECT_SUCCESS(s2n_connection_wipe(client_conn));
+            EXPECT_SUCCESS(s2n_connection_wipe(server_conn));
+            EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+            bool do_tls13 = (i % 2 == 0);
+
+            uint8_t expected_version = 0;
+            if (do_tls13) {
+                expected_version = S2N_TLS13;
+                EXPECT_SUCCESS(s2n_connection_set_config(client_conn, tls13_client_config));
+                EXPECT_SUCCESS(s2n_connection_set_session(client_conn, tls13_ticket.blob.data,
+                        s2n_stuffer_data_available(&tls13_ticket)));
+            } else {
+                expected_version = S2N_TLS12;
+                EXPECT_SUCCESS(s2n_connection_set_config(client_conn, tls12_client_config));
+                EXPECT_SUCCESS(s2n_connection_set_session(client_conn, tls12_ticket.blob.data,
+                        s2n_stuffer_data_available(&tls12_ticket)));
+            }
+
+            EXPECT_OK(s2n_test_negotiate(server_conn, client_conn, &no_early_data));
+            EXPECT_FALSE(ARE_FULL_HANDSHAKES(client_conn, server_conn));
```

Comment:
We could use: `EXPECT_TRUE(ARE_RESUMPTION_HANDSHAKES(client_conn, server_conn));` here to assert the handshakes are indeed resumption type. 

```
#define  ARE_RESUMPTION_HANDSHAKES(client, server) \
    (IS_RESUMPTION_HANDSHAKE(client) && IS_RESUMPTION_HANDSHAKE(server))
```

Reply:
I don't really see the benefit to adding another enum. The only real difference between IS_FULL_HANDSHAKE and IS_RESUMPTION_HANDSHAKE is that we check NEGOTIATED, which I'm not particularly worried about since we just successfully called s2n_negotiate.